research/src/pages/popup/popup.proposal.mdx

```diff
@@ -0,0 +1,342 @@
+[interactive elements](https://html.spec.whatwg.org/multipage/interactive-elements.html):
+
+- `button`
+- `input` in the `button` state
+- `input` in the `text`, `email`, `phone`, or `url` states
+
+Let _activating element_ be the element on which the `popup` attribute is specified. If the
+_activating element_ is a `button` element or an `input` element in the `button` state, and the
+_activating element_ is activated; OR the _activating element_ is an input in the `text`, `email`,
+`phone`, or `url` states and focus is set to the _activating element_, the user agent must run the
+following steps:
+
+1. Let _subject_ be the element with the ID referenced by the `popup` attribute specified on the
+   _activating element_. If the _subject_ is not a `popup` element, then return.
+2. If the `open` IDL attribute on the _subject_ is `true`, run
+   [hiding a `popup` element steps](#hiding-current-popups).
+3. Otherwise, run [showing a `popup` element steps](#showing-popup-steps).
+
+<p class="todo">
+  Probably need to specify some way of storing the relationship once subject is activated, for
+  purposes of deciding whether to hide a popup.
+</p>
+
+<p class="question">
+  Should we have a way to suppress these steps upon focus, so that the author can enable the popup
+  to be shown upon text entry (or some other logic) instead?
+</p>
+
+<h3 id="showing-popup-steps">
+  Showing a <code>popup</code> element steps
+</h3>
+
+1. Run the [hiding currently-shown `popup` elements steps](#hiding-current-popups).
+2. Add the _subject_ to _subject_'s
+   [node document](https://dom.spec.whatwg.org/#concept-node-document)'s
+   [top layer](https://fullscreen.spec.whatwg.org/#top-layer), at the top of the stack.
+3. Set the `open` IDL attribute on the _subject_ to `true`.
+
+<h4 id="hiding-current-popups">
+  Hiding currently-shown <code>popup</code> elements steps
+</h4>
+
+Let _candidate subject_ be the new candidate `popup` element to be shown.
+If applicable, let _activating element_ be the element whose `popup` attribute specifies the ID of
+the _candidate subject_, and which has been activated by the user.
+If applicable, let the _anchor element_ be the element whose ID is referenced by the `anchor`
+specified on the _candidate subject_.
+
+Starting at the bottom of the top-layer stack: for each `popup` element which has an `open` IDL
+attribute set to `true`, and which is not the _candidate subject_:
+
+1. Let _subject_ be the current `popup` element in the top-layer stack.
+2. Walk the DOM to determine if the _subject_ is an ancestor to the _candidate subject_, the
+   _activating element_, or the _anchor element_.
+3. If the _subject_ is an ancestor, then return.
+4. Otherwise, for this _subject_ and any other `popup`s higher in the stack, run the
+   [hiding a `popup` element steps](#hiding-current-popups).
+
+<p class="todo">
+  Step #2 probably needs to be more precise, and more specific about Shadow/light DOM
```

Comment:
Step 2 should definitely refer to the "flat tree" (defined in CSS [here](https://drafts.csswg.org/css-scoping/#flattening)). See Chromium code [here](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/html_popup_element.cc;l=265;drc=1de0bd095828489ebfef0c265e172af3d1dd0afd). This was important to enable the `<selectmenu>`'s use of shadow DOM.